test/async-action.ts

```diff
@@ -214,6 +214,192 @@ test("it should support logging", t => {
     }, 10)
 })
 
+test("it should support asyncActionWithCancel(cb, type) as decorator for have-value variable", t => {
+    const values: any[] = []
+
+    mobx.useStrict(true)
+
+    class X {
+        @mobx.observable a = 1;
+
+        @utils.asyncActionWithCancel(() => {}, 'takeLatest')
+        *f(initial: number) {
+            this.a = initial // this runs in action
+            try {
+                this.a = yield delay(100, 5, true) // and this as well!
+                yield delay(100, 0)
+                this.a = 4
+            } catch (e) {
+                this.a = e
+            }
+            return this.a
+        }
+    }
+
+    const x = new X()
+    mobx.reaction(() => x.a, v => values.push(v), { fireImmediately: true })
+
+    setTimeout(() => {
+        // TODO: mweh on any cast...
+        ;(x.f(/*test binding*/ 2) as any).then((v: number) => {
+            // note: ideally, type of v should be inferred..
+            t.is(v, 5)
+            t.deepEqual(values, [1, 2, 5])
+            t.is(x.a, 5) // correct instance modified?
+            t.end()
+        })
+    }, 10)
+})
+
+test("it should support asyncActionWithCancel(cb, type) as decorator for no-value variable", t => {
+    const values: any[] = []
+
+    mobx.useStrict(true)
+
+    class X {
+        @mobx.observable a = 1;
+
+        @utils.asyncActionWithCancel(() => {}, 'takeLatest')
+        noValueParam
+    }
+
+    const x = new X()
+    x.noValueParam = function *(initial: number) {
+        this.a = initial // this runs in action
+        try {
+            this.a = yield delay(100, 5, true) // and this as well!
+            yield delay(100, 0)
+            this.a = 4
+        } catch (e) {
+            this.a = e
+        }
+        return this.a
+    }
+
+    mobx.reaction(() => x.a, v => values.push(v), { fireImmediately: true })
+
+    setTimeout(() => {
+        // TODO: mweh on any cast...
+        ;(x.noValueParam(/*test binding*/ 2) as any).then((v: number) => {
+            // note: ideally, type of v should be inferred..
+            t.is(v, 5)
+            t.deepEqual(values, [1, 2, 5])
+            t.is(x.a, 5) // correct instance modified?
+            t.end()
+        })
+    }, 10)
+})
+
+test("it should support canceling the effects when call canceler", t => {
+    mobx.useStrict(true)
+    const values: any[] = []
+    const x = mobx.observable({ a: 1 })
+    mobx.reaction(() => x.a, v => values.push(v), { fireImmediately: true })
+
+    let canceler
+    const f = utils.asyncActionWithCancel(cancel => canceler = cancel)(function*(initial: number) {
```

Comment:
API wise, can we just put a `cancel()` method on the return promise? that would probably be simpler, so that one can do:

```javascript
const promise = invokeSomeAsyncAction
promise.cancel()
```

That would probably also remove the need for a global registry?

Reply:
0. I had thought about this good design, but I worry about the lose of the method since **promise.prototype.then** is always return a new promise. 
```
promise.cancel() // This api is great to cancel just a single action
```
1. The ideas of **takeLatest** and **takeEvery** need a place to hold all the action, so that the canceler is able to cancel the prevAction and other same kind of actions.